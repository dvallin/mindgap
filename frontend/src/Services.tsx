import * as React from 'react'
import { PositionService } from './services/location'
import { GeocodingService } from './services/geocoding'
import { StorageService } from './services/storage'

export interface Services {
  positionService: PositionService
  geocodingService: GeocodingService
  storageService: StorageService
}

export const Services = React.createContext<Services | undefined>(undefined)

export const injectServices = <P extends unknown>(WrappedComponent: React.ComponentType<P & Services>): React.ComponentType<P> => (
  props
) => <Services.Consumer>{(services) => services && <WrappedComponent {...services} {...props} />}</Services.Consumer>

export default Services
