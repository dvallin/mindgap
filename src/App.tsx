import * as React from 'react'
import { Route } from 'react-router-dom'

import RecipePage from './recipes/RecipePage'
import BatchPage from './batches/BatchPage'

export default (): JSX.Element => (
  <div className="container">
    <Route exact path="/recipes/:id?" component={RecipePage} />
    <Route exact path="/batches/:id?" component={BatchPage} />
  </div>
)
