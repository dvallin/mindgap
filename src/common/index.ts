import { DateTime } from 'luxon'

export type Unit = string
export type IsoDate = string
export type IsoDuration = string
export type Id = string

export interface Event {
  date: IsoDate
}

export const event = (): Event => ({ date: new Date().toISOString() })

export interface Note {
  note: string
}

export interface Name {
  name: string
  description: string
}

export const name = (): Name => ({ name: '', description: '' })

export function daysAgo(date: IsoDate): number {
  return Math.floor(Math.abs(DateTime.fromISO(date).diffNow('days').days))
}
