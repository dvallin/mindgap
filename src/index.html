<!DOCTYPE html>
<html lang="en" dir="ltr">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bulma@0.9.1/css/bulma.min.css" />
    <title>Mindgap</title>
  </head>

  <body>
    <div id="root"></div>
  </body>
</html>
