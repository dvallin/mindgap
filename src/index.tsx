import * as React from 'react'
import { render } from 'react-dom'

import { Provider } from 'react-redux'

import { createStore } from './store'
const store = createStore()

import App from './App'
import Services from './Services'
import { WebPositionService } from './services/location'
import { Nominatim } from './services/geocoding'
import { LocalStorageService } from './services/storage'

const services: Services = {
  positionService: new WebPositionService(),
  geocodingService: new Nominatim(),
  storageService: new LocalStorageService(),
}

if ('serviceWorker' in navigator) {
  window.addEventListener('load', () => {
    navigator.serviceWorker
      .register('/service-worker.js')
      .then((registration) => {
        console.log('SW registered: ', registration)
      })
      .catch((registrationError) => {
        console.log('SW registration failed: ', registrationError)
      })
  })
}

render(
  <Services.Provider value={services}>
    <Provider store={store}>
      <App />
    </Provider>
  </Services.Provider>,
  document.getElementById('root')
)
