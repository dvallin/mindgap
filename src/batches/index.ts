import { event, Event, Id, IsoDate, name, Name } from '../common'
import { Step } from '../steps'

export type Action = Event & Step

export interface Batch extends Name, Event {
  recipe?: {
    id: Id
    scale: number
  }
  actions: Action[]
  done?: IsoDate
}

export const batch = (): Batch => ({ ...name(), ...event(), actions: [] })
