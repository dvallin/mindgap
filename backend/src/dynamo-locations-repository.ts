import { DynamoDB } from 'aws-sdk'
import { Async } from 'lazy-space'
import { Location, Entity } from './locations'

export class DynamoLocationsRepository {
  constructor(private readonly dynamo: DynamoDB.DocumentClient) {}

  list(): Async<Record<string, unknown>[]> {
    const params: DynamoDB.DocumentClient.ScanInput = {
      TableName: process.env.DYNAMODB_TABLE || '',
    }
    return Async.of(
      new Promise((resolve, reject) => this.dynamo.scan(params, (error, result) => (error ? reject(error) : resolve(result.Items))))
    )
  }

  put(entity: Entity, location: Location): Async<void> {
    const timestamp = new Date().getTime()
    const params: DynamoDB.DocumentClient.PutItemInput = {
      TableName: process.env.DYNAMODB_TABLE || '',
      Item: {
        id: entity,
        json: JSON.stringify(location),
        checked: false,
        createdAt: timestamp,
        updatedAt: timestamp,
      },
    }
    return Async.of(new Promise((resolve, reject) => this.dynamo.put(params, (error) => (error ? reject(error) : resolve()))))
  }
}
