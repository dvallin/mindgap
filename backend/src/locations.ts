import { APIGatewayProxyHandlerV2, APIGatewayProxyStructuredResultV2 } from 'aws-lambda'
import { DynamoDB } from 'aws-sdk'
import { DynamoLocationsRepository } from './dynamo-locations-repository'

export type Entity = string

export interface GeoCoordinates {
  latitude: number
  longitude: number
}

export interface Location {
  coordinates: GeoCoordinates
  data: string
}

export interface LocationComponent {
  entity: Entity
  location: Location
}

const repo = new DynamoLocationsRepository(new DynamoDB.DocumentClient())

export const create: APIGatewayProxyHandlerV2 = async (event) => {
  const data = JSON.parse(event.body || '{}') as LocationComponent
  const result = await repo.put(data.entity, data.location).run()
  return result
    .map<APIGatewayProxyStructuredResultV2>(() => ({
      statusCode: 201,
    }))
    .recover((e) => {
      console.error(e)
      return {
        statusCode: 501,
        headers: { 'Content-Type': 'text/plain' },
        body: "Couldn't create the location item.",
      }
    })
}

export const list: APIGatewayProxyHandlerV2 = async () => {
  const result = await repo.list().run()
  return result
    .map<APIGatewayProxyStructuredResultV2>((items) => ({
      statusCode: 200,
      body: JSON.stringify(items),
    }))
    .recover((e) => {
      console.error(e)
      return {
        statusCode: 501,
        headers: { 'Content-Type': 'text/plain' },
        body: "Couldn't fetch the locations.",
      }
    })
}
