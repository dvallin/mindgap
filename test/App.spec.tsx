import { createShallowMount, mockRouteComponentProps } from './react-test-utils'
import { testBatch, testDate, testRecipe, testState } from './test-utils'

import { App, stateToProps } from '../src/App'
import produce from 'immer'

describe('App', () => {
  const mount = createShallowMount(App, { batches: 3, batchesDone: 2, recipes: 1, ...mockRouteComponentProps({}) })

  describe('Component', () => {
    it('mounts', () => {
      expect(mount()).toMatchSnapshot()
    })
  })

  describe('Store Connection', () => {
    const state = produce(testState, (s) => {
      s.batches.batchCache = {
        batch: testBatch,
        batch2: testBatch,
        done: produce(testBatch, (b) => {
          b.done = testDate
        }),
      }
      s.recipes.recipeCache = { recipe: testRecipe, recipe2: testRecipe }
    })

    const props = stateToProps(state)

    it('extracts counts', () => {
      expect(props).toEqual(
        expect.objectContaining({
          batches: 3,
          batchesDone: 1,
          recipes: 2,
        })
      )
    })
  })
})
