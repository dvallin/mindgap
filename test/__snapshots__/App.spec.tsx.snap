// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component mounts 1`] = `
<div
  className="container"
>
  <section
    className="section"
  >
    <nav
      className="level"
    >
      <div
        className="level-item has-text-centered"
      >
        <NavLink
          activeClassName="is-active"
          exact={true}
          to="/batches?done=false"
        >
          <p
            className="heading"
          >
            Active
          </p>
          <p
            className="title"
          >
            1
          </p>
        </NavLink>
      </div>
      <div
        className="level-item has-text-centered"
      >
        <NavLink
          activeClassName="is-active"
          exact={true}
          to="/batches?done=true"
        >
          <p
            className="heading"
          >
            Done
          </p>
          <p
            className="title"
          >
            2
          </p>
        </NavLink>
      </div>
      <p
        className="level-item has-text-centered"
      >
        <NavLink
          activeClassName="is-active"
          exact={true}
          to="/"
        >
          <img
            alt="Mindgap"
            style={
              Object {
                "height": "60px",
              }
            }
          />
        </NavLink>
      </p>
      <div
        className="level-item has-text-centered"
      >
        <NavLink
          activeClassName="is-active"
          exact={true}
          to="/batches"
        >
          <p
            className="heading"
          >
            Batches
          </p>
          <p
            className="title"
          >
            3
          </p>
        </NavLink>
      </div>
      <div
        className="level-item has-text-centered"
      >
        <NavLink
          activeClassName="is-active"
          exact={true}
          to="/recipes"
        >
          <p
            className="heading"
          >
            Recipes
          </p>
          <p
            className="title"
          >
            1
          </p>
        </NavLink>
      </div>
    </nav>
  </section>
  <Route
    component={[Function]}
    exact={true}
    path="/"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/batches"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/recipes"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/recipes/:id?"
  />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "type": [Function],
      }
    }
    exact={true}
    path="/batches/:id?"
  />
</div>
`;
